overlord/exportstate/exportstate.go

```diff
@@ -0,0 +1,450 @@
+type Manifest struct {
+	Symlinks []SymlinkExport `json:"symlinks,omitempty"`
+}
+
+// NewManifest returns the manifest of a given snap.
+func NewManifest(info *snap.Info) *Manifest {
+	return NewAbstractManifest(info).Materialize()
+}
+
+// IsEmpty returns true if a manifest describes no content.
+func (m *Manifest) IsEmpty() bool {
+	return len(m.Symlinks) == 0
+}
+
+// CreateExportedFiles creates all the files constituting the export manifest.
+//
+// The directory /var/lib/snapd/export/$primaryKey/$subKey is created
+// if necessary. For each export set in the manifest, additional sub-directory
+// is created and populated with symbolic links pointing to the exported files.
+//
+// The function is idempotent.
+func (m *Manifest) CreateExportedFiles() error {
+	for _, s := range m.Symlinks {
+		if err := s.Create(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RemoveExportedFiles removes all the files constituting the export state.
+//
+// In addition the path /var/lib/snapd/export/$primaryKey/$subKey
+// is pruned, removing empty directories if possible.
+//
+// On failure removal continues and the first error is returned.
+func (m *Manifest) RemoveExportedFiles() error {
+	var firstErr error
+	for _, s := range m.Symlinks {
+		if err := s.Remove(); err != nil && firstErr == nil {
+			firstErr = err
+		}
+	}
+	return firstErr
+}
+
+// SymlinkExport describes content exported as symbolic link.
+type SymlinkExport struct {
+	PrimaryKey string `json:"primary-key"`
+	SubKey     string `json:"sub-key"`
+	ExportSet  string `json:"export-set"`
+	Name       string `json:"name"`
+	Target     string `json:"target"`
+}
+
+// PathName returns the full path of the symbolic link.
+func (s *SymlinkExport) PathName() string {
+	return filepath.Join(ExportDir, s.PrimaryKey, s.SubKey, s.ExportSet, s.Name)
+}
+
+// Create creates a symbolic link and necessary directories.
+//
+// The function is idempotent.
+func (s *SymlinkExport) Create() error {
+	pathName := s.PathName()
+	if err := os.MkdirAll(filepath.Dir(pathName), 0755); err != nil {
+		return err
+	}
+	err := os.Symlink(s.Target, pathName)
+	if err != nil && os.IsExist(err) {
+		if actualTarget, _ := os.Readlink(pathName); actualTarget == s.Target {
+			err = nil
+		}
+	}
+	return err
+}
+
+// Remove removes the symbolic link and prunes any directories.
+func (s *SymlinkExport) Remove() error {
+	if err := os.Remove(s.PathName()); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	// XXX: or iterate upwards until we reach ExportDir
+	os.Remove(filepath.Join(ExportDir, s.PrimaryKey, s.SubKey, s.ExportSet))
+	os.Remove(filepath.Join(ExportDir, s.PrimaryKey, s.SubKey))
+	os.Remove(filepath.Join(ExportDir, s.PrimaryKey))
+	return nil
+}
+
+// stateMapKey returns key used for indexing the map of exported snap content.
+func stateMapKey(instanceName string, rev snap.Revision) string {
+	return instanceName + "/" + rev.String()
```

Comment:
Mostly an observation, but organizing the map like this (name/rev as a key) seems a bit unusual, any reason not to map by snap name, then nest revision map inside? No particular concern here, just consistency.

Reply:
Because this allows us to optimize serialization, similarly to what we do in snapstate.